src/helpers/messaging.rs

```diff
@@ -175,13 +181,22 @@ impl Gateway {
                         tracing::trace!("received {} bytes from {:?}", messages.len(), channel_id);
                         receive_buf.receive_messages(&channel_id, &messages);
                     }
-                    Some((channel_id, msg)) = envelope_rx.recv() => {
+                    Some((channel_id, msg, status_tx)) = envelope_rx.recv() => {
                         tracing::trace!("new SendRequest({channel_id:?}, {:?}", msg.record_id);
-                        if let Some(buf_to_send) = send_buf.push(&channel_id, &msg).expect("Failed to append data to the send buffer") {
-                            tracing::trace!("sending {} bytes to {:?}", buf_to_send.len(), &channel_id);
-                            network_sink.send((channel_id, buf_to_send)).await
-                                .expect("Failed to send data to the network");
-                        }
+                        let status = match send_buf.push(&channel_id, &msg) {
+                            Ok(Some(buf_to_send)) => {
+                                tracing::trace!("sending {} bytes to {:?}", buf_to_send.len(), &channel_id);
+                                network_sink.send((channel_id.clone(), buf_to_send)).await
+                                    .expect("Failed to send data to the network");
+                                None
+                            },
+                            Ok(None) => None,
+                            Err(err) => Some(err),
+                        };
+
+                        status_tx.send(status).unwrap_or_else(|_| {
+                            tracing::warn!("No listener for send request {channel_id:?}/{:?}", msg.record_id);
+                        });
```

Comment:
This is now big enough to get its own function.

Reply:
true, moved it to a separate function. One day I would create a struct for this control loop